### PR TITLE
build: specify the key to use when invoking gpg:sign-and-deploy-file

### LINKED ISSUE #16433: mobile: 1.8.3 stuck in deployment pipeline
1.8.3-SNAPSHOT made it to maven central - unfortunately 1.8.3 did not make it - last release version is 1.8.2: http://search.maven.org/#search%7Cga%7C1%7Cgeth

### PR DESCRIPTION
**Pull request by ligi**
This way I could locally upload after the initial problem that the wrong key was used to sign.

This should fix #16433

### CODE REVIEW

**Comment by karalabe** [comment]
I don't really like hard coding credentials into build scripts. The key content is taken from an ENV var. Lets generate the key ID based on the env var, not hard code it into the script. This will just bite us later.

**Comment by ligi** [comment]
done as requested

**Comment by fjl** [comment]
You can use this function to get the key ID with the Go openpgp library.

```go
// PGPKeyID parses an armored key and returns the key ID.
func PGPKeyID(pgpkey string) (string, error) {
	keys, err := openpgp.ReadArmoredKeyRing(bytes.NewBufferString(pgpkey))
	if err != nil {
		return "", err
	}
	if len(keys) != 1 {
		return "", fmt.Errorf("key count mismatch: have %d, want %d", len(keys), 1)
	}
	return keys[0].PrimaryKey.KeyIdString(), nil
}
```

File: build/ci.go [hunk 1]
```diff
@@ -763,15 +763,25 @@ func doAndroidArchive(cmdline []string) {
...
 			gpg.Stdin = bytes.NewReader(key)
 			build.MustRun(gpg)
+
+			keyID, err := build.PGPKeyID(string(key))
+
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			// Upload the artifacts to Sonatype and/or Maven Central
```
Comment (fjl) [inline build/ci.go]: With this change, `-signer` is required for `-upload`. Please remove the `if` around the variable to make that more clear.

File: build/ci.go [hunk 2]
```diff
@@ -755,22 +755,29 @@ func doAndroidArchive(cmdline []string) {
...
+		key, err := base64.StdEncoding.DecodeString(b64key)
+		if err != nil {
+			log.Fatalf("invalid base64 %s", *signer)
 		}
+		gpg := exec.Command("gpg", "--import")
+		gpg.Stdin = bytes.NewReader(key)
+		build.MustRun(gpg)
+
+		keyID, err := build.PGPKeyID(string(key))
+
```
Comment (karalabe) [inline build/ci.go]: Imma be an arse and ask that you delete this empty line. There's absolutely no value it it and it makes the code fall apart. Also please delete the empty line 3 lines below after the closing brace (you already have an entire line almost empty that only contains a brace, that's visually delimited enough not to require extra empty lines).

**Comment by ligi** [comment]
@karalabe done - I usually like to insert some blanks so things don't feel cramped - but obey your will here ;-)

**Comment by Amorn-Kingthong** [comment]
I WILL JOIN

**Comment by ligi** [comment]
@Amorn-Kingthong join what?

**Pull request merged**